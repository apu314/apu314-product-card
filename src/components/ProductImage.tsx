import type { CSSProperties, FC, SyntheticEvent } from "react"

import React, { useContext } from "react"

import { ProductContext } from "./ProductCard"

import noImage from '../assets/no-image.jpg'

import styles from '../styles/styles.module.css'


export interface Props {
  img?: string
  alt?: string
  className?: string
  style?: CSSProperties
}

export const ProductImage: FC<Props> = ({ img = '', alt = '', className, style }) => {

  const { product } = useContext(ProductContext)
  let imgToShow: string

  if (img) {
    imgToShow = img
  } else if (product.img) {
    imgToShow = product.img
  } else {
    imgToShow = noImage
  }

  const onErrorImage = (e: SyntheticEvent<HTMLImageElement, Event>) => e.currentTarget.src = noImage

  return (
    <img
      className={`${styles.productImg} ${className}`}
      style={style}
      src={imgToShow}
      onError={onErrorImage}
      alt={alt}
    />
  )
}
