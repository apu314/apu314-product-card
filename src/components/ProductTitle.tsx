import type { CSSProperties, FC } from "react"

import React, { useContext } from "react"

import { ProductContext } from "./ProductCard"

import styles from '../styles/styles.module.css'


export interface Props {
  title?: string
  className?: string
  style?: CSSProperties
}

export const ProductTitle: FC<Props> = ({ title = '', className, style }) => {
  const { product } = useContext(ProductContext)

  return (
    <span
      className={`${styles.productDescription} ${className}`}
      style={style}
    >
      {title ? title : product.title}
    </span>
  )
}
