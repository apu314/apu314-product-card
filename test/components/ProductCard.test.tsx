import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product1, product2 } from '../data/products';

const { act } = renderer;

describe('ProductCard', () => {
  test('should Show the ProductImage component with the custom image', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => <h1>Product Card</h1>}
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should increment counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {({ count, increaseBy }) => (
          <>
            <h1>Product Card</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(2)}></button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    // Act es para que el test espere a que el estado cambie.
    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();
    expect((tree as any).children[1].children[0]).toBe('2');
  });
});
