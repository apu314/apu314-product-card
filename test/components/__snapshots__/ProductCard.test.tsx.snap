// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should Show the ProductImage component with the custom image 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
</div>
`;

exports[`ProductCard should increment counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;
